### PR TITLE
Add default code to alarm_control_panel

### PR DESCRIPTION
**Pull request by gjohansson-ST**
## Proposed change

Adds capabilities to add a default code to `alarm_control_panel` entities

Core PR: https://github.com/home-assistant/core/pull/112540
Architecture decision: https://github.com/home-assistant/architecture/discussions/892

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

## Summary by CodeRabbit

- **New Features**
  - Introduced `default_code` property for alarm control panel entities, enhancing security and user control.
  - Added functionality to manage alarm modes with code protection, simplifying the process of setting alarm modes.

- **Improvements**
  - Enhanced error handling and subscription management for alarm panel cards, ensuring a more robust user experience.
  - Streamlined the disarming process of alarm control panels with improved logic.

- **Refactor**
  - Simplified code structure by consolidating alarm mode setting logic into a single function, improving maintainability and readability.

### CODE REVIEW

File: src/state-control/alarm_control_panel/ha-state-control-alarm_control_panel-modes.ts
Comment (silamon) [inline src/state-control/alarm_control_panel/ha-state-control-alarm_control_panel-modes.ts:76]:
```suggestion
            `ui.card.alarm_control_panel.${disarm ? "disarm" : "arm"}`
```

File: src/state-control/alarm_control_panel/ha-state-control-alarm_control_panel-modes.ts
Comment (silamon) [inline src/state-control/alarm_control_panel/ha-state-control-alarm_control_panel-modes.ts:67]:
```suggestion
      const alarmControlPanelRegistryEntry =
        await getExtendedEntityRegistryEntry(
          this.hass,
          this.stateObj.entity_id
        ).catch(() => undefined);
      const defaultCode =
        alarmControlPanelRegistryEntry?.options?.alarm_control_panel
          ?.default_code;
```

**Review by silamon — COMMENTED** [review]
Code looks good. 👍 I've did a test and everything works as expected. Just there's a small difference in how the card behaves when a default code is correctly set:
* In the more info dialog, you don't show the prompt anymore to enter the alarm code. 
* In the alarm card panel, you still see the code display while it's not really needed. It should be hidden there, if it's not needed to fill it in.

**Comment by gjohansson-ST** [comment]
> Code looks good. 👍 I've did a test and everything works as expected. Just there's a small difference in how the card behaves when a default code is correctly set:

I did an effort to try to figure it out but I hope maybe to get some help here.
`getExtendedEntityRegistryEntry()` is continuously loading and I think it's probably only worth to set it once when it renders first time. Any help you can offer @silamon ? Thanks

**Comment by silamon** [comment]
Either implement the firstUpdated or connectedCallback in the lit component: 

```
 public async connectedCallback(): Promise<void> {
    super.connectedCallback();
    await this._loadEntityRegistryEntry();
 }
```
  
```
public async firstUpdated(changedProps: PropertyValues) {
    super.firstUpdated(changedProps);
    await this._loadEntityRegistryEntry();
}  
```

I would probably suggest to do connectedCallback, since firstUpdated is called right after the render and obviously would trigger a new render.

**Comment by gjohansson-ST** [comment]
Thanks for the help @silamon
I had to go with `firstUpdated` as `hass` isn't available on the first pass so it seems I need that second render to get it and thereby being able to read the registry.

**Comment by gjohansson-ST** [comment]
Core PR approved

**Comment by silamon** [comment]
Still a few things to decide on in my opinion: 
* While the keypad is gone now, most of the card is also gone.. I think it will need a UX review before approval
image
* When you entered a default code in the settings, you need to refresh the dashboard (note that we don't have an event now to catch an updated entry in a lovelace card)
* When you have a wrong default code and click an alarm card panel action, it shows "invalid alarm code provided". In this case the user didn't provide anything and it might be unclear that it was entered it the settings.

**Comment by silamon** [comment]
I'm also adding the label back to indicate we should "wait for backend" before we can merge.

**Comment by gjohansson-ST** [comment]
> * While the keypad is gone now, most of the card is also gone.. I think it will need a UX review before approval

I guess that makes sense. The card is a bit odd without anything really in it.

> * When you entered a default code in the settings, you need to refresh the dashboard (note that we don't have an event now to catch an updated entry in a lovelace card)

Can you provide some guidance how to fix that. As I have no clue otherwise.

> * When you have a wrong default code and click an alarm card panel action, it shows "invalid alarm code provided". In this case the user didn't provide anything and it might be unclear that it was entered it the settings.

Isn't that an error coming from the respective integration?

**Comment by piitaya** [comment]
> Still a few things to decide on in my opinion:
> 
> * While the keypad is gone now, most of the card is also gone.. I think it will need a UX review before approval

The design is ok. User can use other card if they don't have code (e.g. tile card)
> * When you entered a default code in the settings, you need to refresh the dashboard (note that we don't have an event now to catch an updated entry in a lovelace card)

We should be able to detect the changes using this code: 

```ts
protected updated(changedProps: PropertyValues<typeof this>): void {
    super.updated(changedProps);

    if (changedProps.has("hass")) {
      const oldHass = changedProps.get("hass");
      const entityId = this._config?.entity;
      if (entityId) {
        if (oldHass?.entities[entityId] !== this.hass?.entities[entityId]) {
          this._loadEntityRegistryEntry();
        }
      }
    }
  }
  ```

**Comment by silamon** [comment]
> > * When you have a wrong default code and click an alarm card panel action, it shows "invalid alarm code provided". In this case the user didn't provide anything and it might be unclear that it was entered it the settings.
> 
> Isn't that an error coming from the respective integration?

I just checked and yes, it's an error coming from the implementing integration. 
Other questions have been answered by Paul. (Thank you, Paul)

**Comment by silamon** [comment]
Retested, but it keeps reloading the registry again right now, eventually slowing down the frontend.

**Comment by gjohansson-ST** [comment]
> Retested, but it keeps reloading the registry again right now, eventually slowing down the frontend.

I changed it again but I guess you tested it before.
With @piitaya suggestion it doesn't work as there is nothing triggering the `update` function at the moment.
I guess I need to modify `shouldUpdate` somehow but not sure how to only update when the default code is updated.

**Comment by silamon** [comment]
> > Retested, but it keeps reloading the registry again right now, eventually slowing down the frontend.
> 
> I changed it again but I guess you tested it before. With @piitaya suggestion it doesn't work as there is nothing triggering the `update` function at the moment. I guess I need to modify `shouldUpdate` somehow but not sure how to only update when the default code is updated.

No, I see. There's nothing that triggers the update and it's impossible from the more-info dialog to retrigger a card update at this time. Using a subscription on the entity does the trick, but I'm unsure if it would be accepted and the most performable solution for this case:  

```
diff --git a/src/panels/lovelace/cards/hui-alarm-panel-card.ts b/src/panels/lovelace/cards/hui-alarm-panel-card.ts
index c5b11775b..e29cd5684 100644
--- a/src/panels/lovelace/cards/hui-alarm-panel-card.ts
+++ b/src/panels/lovelace/cards/hui-alarm-panel-card.ts
@@ -1,4 +1,4 @@
-import { HassEntity } from "home-assistant-js-websocket";
+import { HassEntity, UnsubscribeFunc } from "home-assistant-js-websocket";
 import {
   CSSResultGroup,
   LitElement,
@@ -33,6 +33,7 @@ import type { LovelaceCard } from "../types";
 import {
   ExtEntityRegistryEntry,
   getExtendedEntityRegistryEntry,
+  subscribeEntityRegistry,
 } from "../../../data/entity_registry";
 import { AlarmPanelCardConfig, AlarmPanelCardConfigState } from "./types";
 
@@ -107,6 +108,20 @@ class HuiAlarmPanelCard extends LitElement implements LovelaceCard {
 
   @query("#alarmCode") private _input?: HaTextField;
 
+  private _unsubEntityRegistry?: UnsubscribeFunc;
+
+  public connectedCallback() {
+    super.connectedCallback();
+  }
+
+  public disconnectedCallback() {
+    super.disconnectedCallback();
+    if (this._unsubEntityRegistry) {
+      this._unsubEntityRegistry();
+      this._unsubEntityRegistry = undefined;
+    }
+  }
+
   public async getCardSize(): Promise<number> {
     if (!this._config || !this.hass) {
       return 9;
@@ -136,18 +151,12 @@ class HuiAlarmPanelCard extends LitElement implements LovelaceCard {
     if (!this._config || !this.hass) {
       return;
     }
+
     const oldHass = changedProps.get("hass") as HomeAssistant | undefined;
     const oldConfig = changedProps.get("_config") as
       | AlarmPanelCardConfig
       | undefined;
-    if (changedProps.has("hass")) {
-      const entityId = this._config?.entity;
-      if (entityId) {
-        if (oldHass?.entities[entityId] !== this.hass?.entities[entityId]) {
-          this._loadEntityRegistryEntry();
-        }
-      }
-    }
+
     if (
       !oldHass ||
       !oldConfig ||
@@ -183,9 +192,18 @@ class HuiAlarmPanelCard extends LitElement implements LovelaceCard {
       return;
     }
     try {
-      this._entry = await getExtendedEntityRegistryEntry(
-        this.hass!,
-        this._config!.entity
+      this._unsubEntityRegistry = subscribeEntityRegistry(
+        this.hass!.connection,
+        async (entries) => {
+          if (
+            entries.some((entry) => entry.entity_id === this._config!.entity)
+          ) {
+            this._entry = await getExtendedEntityRegistryEntry(
+              this.hass!,
+              this._config!.entity
+            );
+          }
+        }
       );
     } catch (e) {
       this._entry = null;
```

**Comment by gjohansson-ST** [comment]
> No, I see. There's nothing that triggers the update and it's impossible from the more-info dialog to retrigger a card update at this time. Using a subscription on the entity does the trick, but I'm unsure if it would be accepted and the most performable solution for this case:

I adapted the code to your proposal and it works like a charm.
I noticed it's only the area card which already has this subscription but then it sort of confirms it's the only only way to handle it right now.

**Comment by piitaya** [comment]
@gjohansson-ST I updated the PR to have it ready for today's beta.
Here's the fixes : 
- check for default code for every alarm call (It was missing for card features and disarm in more info)
- fix subscription to entity registry

**Review by piitaya — APPROVED** [review]

**Comment by gjohansson-ST** [comment]
> @gjohansson-ST I updated the PR to have it ready for today's beta. Here's the fixes :

Thanks @piitaya 👍

**Pull request merged**
